src/GafferScene/CameraQuery.cpp

```diff
@@ -0,0 +1,551 @@
+		if( queryPlug == nullptr )
+		{
+			throw IECore::Exception( "CameraQuery::affects : Query plugs must be \"StringPlug\"" );
+		}
+
+		const ValuePlug *valuePlug = valuePlugFromQuery( queryPlug );
+
+		addChildPlugsToAffectedOutputs( valuePlug, outputs );
+
+		outputs.push_back( sourcePlugFromQuery( queryPlug ) );
+	}
+}
+
+void CameraQuery::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	if( output == internalParametersPlug() )
+	{
+		std::string location = "";
+		// The globals affect the camera parameters via SceneAlgo::applyCameraGlobals()
+		h.append( scenePlug()->globalsHash() );
+
+		h.append( cameraModePlug()->getValue() );
+		if( cameraModePlug()->getValue() == (int)CameraMode::RenderCamera )
+		{
+			if( const auto renderCameraData = scenePlug()->globals()->member<StringData>( g_renderCameraOption ) )
+			{
+				location = renderCameraData->readable();
+			}
+		}
+		else
+		{
+			location = locationPlug()->getValue();
+		}
+
+		if( !location.empty() )
+		{
+			const ScenePlug::ScenePath locationPath = ScenePlug::stringToPath( location );
+			ScenePlug::PathScope scope( context, &locationPath );
+			if( scenePlug()->existsPlug()->getValue() )
+			{
+				h.append( scenePlug()->objectHash( locationPath ) );
+			}
+		}
+	}
+	else if( outPlug()->isAncestorOf( output ) )
+	{
+		const ValuePlug *oPlug = outPlug( output );
+
+		if( output == oPlug->getChild( g_sourcePlugIndex ) )
+		{
+			internalParametersPlug()->hash( h );
+			const StringPlug *childQueryPlug = queryPlug( output );
+			childQueryPlug->hash( h );
+		}
+		else if(
+			oPlug->getChild( g_valuePlugIndex )->isAncestorOf( output ) ||
+			output == oPlug->getChild( g_valuePlugIndex )
+		)
+		{
+			internalParametersPlug()->hash( h );
+			const StringPlug *childQueryPlug = queryPlug( output );
+			childQueryPlug->hash( h );
+		}
+	}
+
+	ComputeNode::hash( output, context, h );
+}
+
+void CameraQuery::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
+{
+	if( output == internalParametersPlug() )
+	{
+		IECore::CompoundDataPtr parameters = new IECore::CompoundData;
+
+		std::string location = "";
+		if( cameraModePlug()->getValue() == (int)CameraMode::RenderCamera )
+		{
+			if( const auto cameraData = scenePlug()->globals()->member<StringData>( g_renderCameraOption ) )
+			{
+				location = cameraData->readable();
+			}
+		}
+		else
+		{
+			location = locationPlug()->getValue();
+		}
+
+		if( !location.empty() )
+		{
+			const ScenePlug::ScenePath locationPath = ScenePlug::stringToPath( location );
+			ScenePlug::PathScope scope( context, &locationPath );
+			if( scenePlug()->existsPlug()->getValue() )
+			{
+				if( const auto camera = runTimeCast<const IECoreScene::Camera>( scenePlug()->objectPlug()->getValue() ) )
+				{
+					auto cameraWithGlobals = camera->copy();
+					SceneAlgo::applyCameraGlobals( cameraWithGlobals.get(), scenePlug()->globals().get(), scenePlug() );
```

Comment:
We use `globals()` again below - it would be worth calling just once and assigning to a variable I think.